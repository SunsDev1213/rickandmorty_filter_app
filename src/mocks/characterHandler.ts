import { rest } from 'msw';

const characterInfo = {
	info: {
		count: '826',
		pages: '42',
		next: 'https://rickandmortyapi.com/api/character?page=2',
		prev: 'null'
	},
	results: [
		{
			id: 1,
			name: 'Rick Sanchez',
			status: 'Alive',
			species: 'Human',
			gender: 'Male'
		},
		{
			id: 2,
			name: 'Morty Smith',
			status: 'Alive',
			species: 'Human',
			gender: 'Male'
		}
	]
};

export const characterHandler = rest.get('https://rickandmortyapi.com/api/character', (_, res, ctx) => {
	return res(ctx.status(200), ctx.json(characterInfo));
});

export const characterDetailHandler = rest.get('https://rickandmortyapi.com/api/character/1', (_, res, ctx) => {
	return res(ctx.status(200), ctx.json(characterInfo));
});
