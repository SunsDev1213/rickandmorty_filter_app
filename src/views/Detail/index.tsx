import React from 'react'
import { useParams, useHistory } from "react-router-dom";
import { Grid, Button } from '@mui/material';
import ChevronLeftIcon from '@mui/icons-material/ChevronLeft';
// Rest API
import { getCharacterById } from '../../hooks';

export default function Detail() {
    let params = useParams();
    let history = useHistory();
    const [characterId, setCharacterId] = React.useState<any>()
    const [characterInfo, setCharacterInfo] = React.useState<any>()

    React.useEffect(() => {
        setCharacterId(params)
    }, [])

    React.useEffect(() => {
        if(characterId?.id) {
            getCharacterById(characterId?.id).then((res:any) => {
                setCharacterInfo(res)
            })
        }
    }, [characterId?.id])  

    const handleBack = () => {
        history.push('/');
    }

    return (
        <div className='root'>
			<div className='container'>
                <Button variant="outlined" onClick={handleBack} startIcon={<ChevronLeftIcon />} style={{marginBottom: 20}}>Back</Button>
                <Grid container spacing={2} >
                    <Grid item xs={12} sm={6}>
                        <img src={characterInfo?.image} alt="avatar" width="100%" className="image"  />
                    </Grid>
                    <Grid item xs={12} sm={6} style={{display: "grid"}}>
                        <div>
                            <span className='desc'>Name: </span>
                            <span className='name'>{characterInfo?.name}</span>
                        </div>
                        <div>
                            <span className='desc'>Status: </span>
                            <span className='name'>{characterInfo?.status}</span>
                        </div>
                        <div>
                            <span className='desc'>Species: </span>
                            <span className='name'>{characterInfo?.species}</span>
                        </div>
                        <div>
                            <span className='desc'>Gender: </span>
                            <span className='name'>{characterInfo?.gender}</span>
                        </div>
                        <div>
                            <span className='desc'>Origin: </span>
                            <span className='name'>{characterInfo?.origin?.name}</span>
                        </div>
                        <div>
                            <span className='desc'>Last known location: </span>
                            <span className='name'>{characterInfo?.location?.name}</span>
                        </div>
                        <div>
                            <span className='desc'>Number of episodes appearances: </span>
                            <span className='name'>{characterInfo?.episode?.length}</span>
                        </div>
                    </Grid>
                </Grid>
            </div>
        </div>
    )
}