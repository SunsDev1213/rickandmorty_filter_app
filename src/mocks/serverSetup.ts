import { setupServer } from 'msw/node';
import { characterHandler, characterDetailHandler } from './characterHandler';

const server = setupServer(characterHandler, characterDetailHandler);

export { server };
