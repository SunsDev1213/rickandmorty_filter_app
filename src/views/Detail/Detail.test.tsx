import '@testing-library/jest-dom';
import { render, screen, act, cleanup, waitFor } from '@testing-library/react';
import Detail from './index';
import Router from 'react-router-dom';
import { server } from '../../mocks/serverSetup';

jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useParams: jest.fn()
}));

const createWrapper = () => {
	return render(<Detail />);
};

beforeAll(() => server.listen());
afterEach(() => {
	cleanup();
	server.resetHandlers();
});
afterAll(() => server.close());

describe('Detail Page', () => {
	test('renders character detail page', async () => {
		jest.spyOn(Router, 'useParams').mockReturnValue({ id: '1' });
		const wrapper = createWrapper();
		expect(wrapper).toMatchSnapshot();
	});
});
