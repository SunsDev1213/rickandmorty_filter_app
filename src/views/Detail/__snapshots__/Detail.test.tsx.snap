// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail Page renders character detail page 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="root"
      >
        <div
          class="container"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
            style="margin-bottom: 20px;"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ChevronLeftIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                />
              </svg>
            </span>
            Back
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-1b3l6lk-MuiGrid-root"
            >
              <img
                alt="avatar"
                class="image"
                width="100%"
              />
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-1b3l6lk-MuiGrid-root"
              style="display: grid;"
            >
              <div>
                <span
                  class="desc"
                >
                  Name: 
                </span>
                <span
                  class="name"
                />
              </div>
              <div>
                <span
                  class="desc"
                >
                  Status: 
                </span>
                <span
                  class="name"
                />
              </div>
              <div>
                <span
                  class="desc"
                >
                  Species: 
                </span>
                <span
                  class="name"
                />
              </div>
              <div>
                <span
                  class="desc"
                >
                  Gender: 
                </span>
                <span
                  class="name"
                />
              </div>
              <div>
                <span
                  class="desc"
                >
                  Origin: 
                </span>
                <span
                  class="name"
                />
              </div>
              <div>
                <span
                  class="desc"
                >
                  Last known location: 
                </span>
                <span
                  class="name"
                />
              </div>
              <div>
                <span
                  class="desc"
                >
                  Number of episodes appearances: 
                </span>
                <span
                  class="name"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="root"
    >
      <div
        class="container"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
          style="margin-bottom: 20px;"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ChevronLeftIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              />
            </svg>
          </span>
          Back
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-1b3l6lk-MuiGrid-root"
          >
            <img
              alt="avatar"
              class="image"
              width="100%"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-1b3l6lk-MuiGrid-root"
            style="display: grid;"
          >
            <div>
              <span
                class="desc"
              >
                Name: 
              </span>
              <span
                class="name"
              />
            </div>
            <div>
              <span
                class="desc"
              >
                Status: 
              </span>
              <span
                class="name"
              />
            </div>
            <div>
              <span
                class="desc"
              >
                Species: 
              </span>
              <span
                class="name"
              />
            </div>
            <div>
              <span
                class="desc"
              >
                Gender: 
              </span>
              <span
                class="name"
              />
            </div>
            <div>
              <span
                class="desc"
              >
                Origin: 
              </span>
              <span
                class="name"
              />
            </div>
            <div>
              <span
                class="desc"
              >
                Last known location: 
              </span>
              <span
                class="name"
              />
            </div>
            <div>
              <span
                class="desc"
              >
                Number of episodes appearances: 
              </span>
              <span
                class="name"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
